Content.Server/Nyanotrasen/Research/Oracle/OracleSystem.cs

```diff
@@ -1,258 +1,288 @@
+            NextItem(oracle);
+    }
 
-        while (i != 0)
-        {
-            EntityManager.SpawnEntity("MaterialBluespace1", Transform(args.User).Coordinates);
-            i--;
-        }
+    private void SendTelepathicInfo(Entity<OracleComponent> oracle, INetChannel client, string message)
+    {
+        var messageWrap = Loc.GetString("chat-manager-send-telepathic-chat-wrap-message",
+            ("telepathicChannelName", Loc.GetString("chat-manager-telepathic-channel-name")),
+            ("message", message));
 
-        if (nextItem)
-            NextItem(component);
+        _chatMan.ChatMessageToOne(ChatChannel.Telepathic,
+            message, messageWrap, oracle, false, client, Color.PaleVioletRed);
     }
 
-    private bool CheckValidity(EntityPrototype given, EntityPrototype target)
+    private bool IsCorrectItem(EntityPrototype given, EntityPrototype target)
     {
-        // 1: directly compare Names
-        // name instead of ID because the oracle asks for them by name
-        // this could potentially lead to like, labeller exploits maybe but so far only mob names can be fully player-set.
+        // Nyano, what is this shit?
+        // Why are we comparing by name instead of prototype id?
+        // Why is this ever necessary?
+        // What were you trying to accomplish?!
         if (given.Name == target.Name)
             return true;
 
         return false;
     }
 
-    private void DispenseLiquidReward(EntityUid uid, OracleComponent component)
+    private void DispenseRewards(Entity<OracleComponent> oracle, EntityCoordinates throwTarget)
     {
-        if (!_solutionSystem.TryGetSolution(uid, OracleComponent.SolutionName, out var fountainSol))
-            return;
+        foreach (var rewardRandom in oracle.Comp.RewardEntities)
+        {
+            // Spawn each reward next to oracle and throw towards the target
+            var rewardProto = _protoMan.Index(rewardRandom).Pick(_random);
+            var reward = EntityManager.SpawnNextToOrDrop(rewardProto, oracle);
+            _throwing.TryThrow(reward, throwTarget, recoil: false);
+        }
 
-        var allReagents = _prototypeManager.EnumeratePrototypes<ReagentPrototype>()
-            .Where(x => !x.Abstract)
-            .Select(x => x.ID).ToList();
+        DispenseLiquidReward(oracle);
+    }
 
-        var amount = 20 + _random.Next(1, 30) + _glimmerSystem.Glimmer / 10f;
-        amount = (float) Math.Round(amount);
+    private void DispenseLiquidReward(Entity<OracleComponent> oracle)
+    {
+        if (!_solutions.TryGetSolution(oracle.Owner, OracleComponent.SolutionName, out var fountainSol))
+            return;
 
-        var sol = new Solution();
-        var reagent = "";
+        // Why is this hardcoded?
+        var amount = MathF.Round(20 + _random.Next(1, 30) + _glimmer.Glimmer / 10f);
+        var temporarySol = new Solution();
+        var reagent = _protoMan.Index(oracle.Comp.RewardReagents).Pick(_random);
```

Comment:
Maybe you could have the oracle be aware of the research tier of the item being requested, and improve the rewards by some factor of research tier? That way science is given better rewards for harder to obtain items?

Reply:
If you can suggest a particular implementation, sure, otherwise introducing new mechanics is not within the scope of this PR.

Reply:
That's fair enough! I personally was planning on making this section a little more variable with a future implementation in refactored psionics, but for now it's good enough.